protocol/x/accountplus/ante/ante.go

```diff
@@ -0,0 +1,299 @@
+	)
+
+	// Make sure smart account is active.
+	if active := ad.accountPlusKeeper.GetIsSmartAccountActive(ctx); !active {
+		return ctx, types.ErrSmartAccountNotActive
+	}
+
+	// Authenticators don't support manually setting the fee payer
+	err = ad.ValidateAuthenticatorFeePayer(tx)
+	if err != nil {
+		return sdk.Context{}, err
+	}
+
+	msgs := tx.GetMsgs()
+	if len(msgs) == 0 {
+		return ctx, errorsmod.Wrap(sdkerrors.ErrInvalidRequest, "no messages in transaction")
+	}
+
+	feeTx, ok := tx.(sdk.FeeTx)
+	if !ok {
+		return ctx, errorsmod.Wrap(sdkerrors.ErrTxDecode, "Tx must be a FeeTx")
+	}
+
+	// The fee payer is the first signer of the transaction. This should have been enforced by the
+	// LimitFeePayerDecorator
+	signers, _, err := ad.cdc.GetMsgV1Signers(msgs[0])
+	if err != nil {
+		return ctx, errorsmod.Wrap(sdkerrors.ErrUnauthorized, "failed to get signers")
+	}
+	feePayer := sdk.AccAddress(signers[0])
+	feeGranter := feeTx.FeeGranter()
+	fee := feeTx.GetFee()
+
+	selectedAuthenticators, err := ad.GetSelectedAuthenticators(tx, len(msgs))
+	if err != nil {
+		return ctx, err
+	}
+
+	// tracks are used to make sure that we only write to the store after every message is successful
+	var tracks []func() error
+
+	// Authenticate the accounts of all messages
+	for msgIndex, msg := range msgs {
+		signers, _, err := ad.cdc.GetMsgV1Signers(msg)
+		if err != nil {
+			return ctx, errorsmod.Wrap(sdkerrors.ErrUnauthorized, "failed to get signers")
+		}
+		// Enforce only one signer per message
+		if len(signers) != 1 {
+			return sdk.Context{}, errorsmod.Wrap(sdkerrors.ErrUnauthorized, "messages must have exactly one signer")
+		}
+
+		// By default, the first signer is the account that is used
+		account := sdk.AccAddress(signers[0])
+
+		// Get the currently selected authenticator
+		selectedAuthenticatorId := selectedAuthenticators[msgIndex]
+		selectedAuthenticator, err := ad.accountPlusKeeper.GetInitializedAuthenticatorForAccount(
+			ctx,
+			account,
+			selectedAuthenticatorId,
+		)
+		if err != nil {
+			return sdk.Context{},
+				errorsmod.Wrapf(
+					err,
+					"failed to get initialized authenticator "+
+						"(account = %s, authenticator id = %d, msg index = %d, msg type url = %s)",
+					account,
+					selectedAuthenticatorId,
+					msgIndex,
+					sdk.MsgTypeURL(msg),
+				)
+		}
+
+		// Generate the authentication request data
+		authenticationRequest, err := authenticator.GenerateAuthenticationRequest(
+			ctx,
+			ad.cdc,
+			ad.accountKeeper,
+			ad.sigModeHandler,
+			account,
+			feePayer,
+			feeGranter,
+			fee,
+			msg,
+			tx,
+			msgIndex,
+			simulate,
+		)
+		if err != nil {
+			return sdk.Context{},
+				errorsmod.Wrapf(
+					err,
+					"failed to generate authentication data "+
+						"(account = %s, authenticator id = %d, msg index = %d, msg type url = %s)",
+					account,
+					selectedAuthenticator.Id,
+					msgIndex,
+					sdk.MsgTypeURL(msg),
+				)
+		}
+
+		a11r := selectedAuthenticator.Authenticator
```

Comment:
Why this variable naming?

Reply:
I guess this variable name was too long for them so they used something similar to a16z.

I don't have a preference on this, let me know if you do.

Reply:
Nit: let's stick to our naming conventions and use sth like `authenticator` or `selectedAuth`?